### PR TITLE
Add Body Battery Support

### PR DESCRIPTION
**Pull request by bes-dev**
Add Body Battery Support

Summary

- Add comprehensive Body Battery metric support to garth library
- Implement both legacy Body Battery events and daily Body Battery with stress data
- Provide complete test coverage following existing patterns

What's Added

New Data Classes

BodyBatteryData - Legacy Body Battery events data
```python
# Get Body Battery events for a specific date (e.g., sleep events)
events = garth.BodyBatteryData.get("2023-07-20")
for event in events:
    if event.event:
        print(f"Event: {event.event.event_type}, Impact: {event.event.body_battery_impact}")
    print(f"Current level: {event.current_level}")
```

DailyBodyBatteryStress - Complete daily Body Battery and stress data
```python
# Get comprehensive Body Battery and stress data for a day
daily_data = garth.DailyBodyBatteryStress.get("2023-07-20")
if daily_data:
    print(f"Current Body Battery: {daily_data.current_body_battery}")
    print(f"Daily change: {daily_data.body_battery_change}")
    print(f"Max stress: {daily_data.max_stress_level}")

    # Access individual readings
    for reading in daily_data.body_battery_readings:
        print(f"Time: {reading.timestamp}, Level: {reading.level}, Status: {reading.status}")
```

Key Features

1. Dual API Support:
  - /wellness-service/wellness/bodyBattery/events/{date} - Event-based data (sleep, activities)
  - /wellness-service/wellness/dailyStress/{date} - Complete daily data with stress
2. Rich Data Properties:
  - current_body_battery, max_body_battery, min_body_battery
  - body_battery_change - Daily Body Battery difference
  - body_battery_readings - Structured timestamp/level data
  - stress_readings - Parallel stress level data
3. Consistent API:
  - .get(date) - Single day data
  - .list(end_date, period) - Multiple days with automatic date handling

API Endpoints Used

- GET /wellness-service/wellness/bodyBattery/events/{date} - Body Battery events
- GET /wellness-service/wellness/dailyStress/{date} - Daily stress + Body Battery data

Test Coverage

- Complete test suite with 6 test functions covering all methods and edge cases
- VCR cassettes for reproducible API testing
- Property validation and error handling tests
- Follows existing test patterns for HRV and Sleep data

Files Changed

- src/garth/data/body_battery.py - New Body Battery implementation
- src/garth/data/__init__.py - Export new classes
- src/garth/__init__.py - Add to public API
- tests/data/test_body_battery_data.py - Comprehensive test suite
- tests/data/cassettes/ - 6 new test cassettes

Example Usage
```python
import garth
from datetime import date

# Authenticate
garth.resume("~/.garth")

# Get current Body Battery data
today_data = garth.DailyBodyBatteryStress.get(date.today())
if today_data:
    print(f"🔋 Current Body Battery: {today_data.current_body_battery}")
    print(f"📊 Daily change: {today_data.body_battery_change}")
    print(f"😰 Average stress: {today_data.avg_stress_level}")

# Get Body Battery events (sleep, activities)
events = garth.BodyBatteryData.get(date.today())
for event in events:
    if event.event and event.event.event_type == "sleep":
        print(f"😴 Sleep impact: +{event.event.body_battery_impact}")

# Get week of data
week_data = garth.DailyBodyBatteryStress.list(period=7)
for day in week_data:
    print(f"{day.calendar_date}: {day.current_body_battery}")
```

This implementation fills a significant gap in the library's health metrics coverage, providing access to one of Garmin's most popular features alongside existing HRV, sleep, stress, and activity data.

## Summary by CodeRabbit

- **New Features**
  - Introduced support for Body Battery and stress data, including daily summaries and event-based details.
  - Added new data structures for Body Battery readings, stress readings, and related metrics.
  - Enabled retrieval and listing of Body Battery and stress data over multiple days.

- **Tests**
  - Added comprehensive tests for Body Battery and stress data retrieval, properties, and error handling.

- **Chores**
  - Updated development dependencies management and reorganized dependency groups in project configuration.
  - Improved installation commands in the Makefile for development, linting, and testing dependencies.

### CODE REVIEW

File: src/garth/data/body_battery.py [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+            event_data = item.get("event")
+            event = None
+            if event_data:
+                event = BodyBatteryEvent(
+                    event_type=event_data.get("eventType", ""),
+                    event_start_time_gmt=datetime.fromisoformat(
+                        event_data.get("eventStartTimeGmt", "").replace("Z", "+00:00")
+                    ),
+                    timezone_offset=event_data.get("timezoneOffset", 0),
+                    duration_in_milliseconds=event_data.get("durationInMilliseconds", 0),
```
[the remark this answers is not in the document]
Comment (bes-dev) [reply src/garth/data/body_battery.py]: done

File: src/garth/data/body_battery.py [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+        all_data = []
+
+        for i in range(period):
+            date_to_fetch = end_date - timedelta(days=i)
+            data = cls.get(date_to_fetch, client=client)
+            if data:
+                all_data.append(data)
+
+        return all_data
+
```
Comment (matin) [inline src/garth/data/body_battery.py:189]: why not use [`Data.list()`](https://github.com/matin/garth/blob/main/src/garth/data/_base.py#L22). it functionally does the same thing but with threads

File: src/garth/data/body_battery.py [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+@dataclass
+class StressReading:
+    """Individual stress reading."""
+    timestamp: int
+    stress_level: int
+
+
+@dataclass
+class DailyBodyBatteryStress:
```
Comment (matin) [inline src/garth/data/body_battery.py:43]: why not inherit from `Data`? that'd give you `.list()` and consistency

File: src/garth/data/body_battery.py [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+
+@dataclass
+class BodyBatteryData:
+    """Legacy Body Battery events data (sleep events only)."""
+    event: BodyBatteryEvent | None = None
+    activity_name: str | None = None
+    activity_type: str | None = None
+    activity_id: str | None = None
+    average_stress: float | None = None
+    stress_values_array: List[List[int]] | None = None
```
Comment (matin) [inline src/garth/data/body_battery.py:199]: please use `list[]` instead of `List[]` for consistency

File: src/garth/data/body_battery.py [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+    body_battery_values_array: List[List[Any]] | None = None
+
+    @property
+    def body_battery_readings(self) -> List[BodyBatteryReading]:
+        """Convert body battery values array to structured readings."""
+        if not self.body_battery_values_array:
+            return []
```
Comment (matin) [inline src/garth/data/body_battery.py:206]: is this necessary? if `self.body_batter_values_array` is empty, the `for` loop would just not run

**Review by matin — CHANGES_REQUESTED** [review]
great functionality to add. thanks for taking the time to work on this.

the main thing I'd change is taking advantage of `camel_to_snake()` and the fact that Pydantic will already manage coerce types like datetimes and handle defaults. that gives you the ability to do `cls(**response)` in many cases, especially because pydantic will apply the logic to custom types for individual fields as ell. those changes would simply your code

please also break up each resource into a different file for clarity

File: src/garth/data/body_battery.py [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+        )
+        end_gmt = datetime.fromisoformat(
+            response.get("endTimestampGMT", "").replace("Z", "+00:00")
+        )
+        start_local = datetime.fromisoformat(
+            response.get("startTimestampLocal", "").replace("Z", "+00:00")
+        )
+        end_local = datetime.fromisoformat(
+            response.get("endTimestampLocal", "").replace("Z", "+00:00")
+        )
```
[the remark this answers is not in the document]
Comment (bes-dev) [reply src/garth/data/body_battery.py]: done

File: src/garth/data/body_battery.py [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+            stress_chart_y_axis_origin=response.get(
+                "stressChartYAxisOrigin", 0
+            ),
+            stress_values_array=response.get(
+                "stressValuesArray", []
+            ),
+            body_battery_values_array=response.get(
+                "bodyBatteryValuesArray", []
+            )
+        )
```
[the remark this answers is not in the document]
Comment (bes-dev) [reply src/garth/data/body_battery.py]: done

File: src/garth/data/body_battery.py (hunk 4 above)
Comment (bes-dev) [reply src/garth/data/body_battery.py]: done

File: src/garth/data/body_battery_events.py [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+                    activity_id=item.get("activityId"),
+                    average_stress=item.get("averageStress"),
+                    stress_values_array=item.get("stressValuesArray"),
+                    body_battery_values_array=item.get(
+                        "bodyBatteryValuesArray"
+                    ),
+                )
+            )
+
+        return events
```
[the remark this answers is not in the document]
Comment (bes-dev) [reply src/garth/data/body_battery_events.py]: done

File: src/garth/data/body_battery/events.py [hunk 9]
```diff
@@ -0,0 +1,149 @@
+from datetime import date, datetime, timedelta
```
[the remark this answers is not in the document]
Comment (bes-dev) [reply src/garth/data/body_battery/events.py]: done

File: src/garth/data/body_battery/events.py [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+    def body_battery_readings(self) -> list[BodyBatteryReading]:
+        """Convert body battery values array to structured readings."""
+        return parse_body_battery_readings(self.body_battery_values_array)
+
+    @property
+    def current_level(self) -> int | None:
+        """Get the latest Body Battery level."""
+        readings = self.body_battery_readings
+        return readings[-1].level if readings else None
+
```
[the remark this answers is not in the document]
Comment (bes-dev) [reply src/garth/data/body_battery/events.py:51]: done

**Comment by bes-dev** [comment]
@matin PR is ready for review

## Architectural Refactoring: Body Battery Module Split

### Summary
Implemented the requested separation of Body Battery resources into dedicated files, following project patterns and addressing technical debt.

### Changes Made

**File Structure:**
data/body_battery/
├── readings.py      # Shared data classes + parsing utilities
├── events.py        # Legacy BodyBatteryData + BodyBatteryEvent├── daily_stress.py  # Modern DailyBodyBatteryStress
└── init.py      # Clean re-exports

**Technical Improvements:**
- ✅ **Separated resources** as requested - events API vs daily stress API
- ✅ **Eliminated circular imports** by extracting shared classes to `readings.py`
- ✅ **Implemented DRY principle** with reusable `parse_*` functions
- ✅ **Added threading** to `BodyBatteryData.list()` for consistency
- ✅ **Applied modern typing** - `list[]` instead of `List[]` throughout
- ✅ **Simplified constructors** - using `camel_to_snake_dict()` + `cls(**response)` pattern

**Comment by matin** [comment]
@bes-dev could you take a look at the failing tasks (lint and tests)?

File: tests/data/test_body_battery_data.py [hunk 11]
```diff
@@ -0,0 +1,287 @@
...
+            reading = readings[0]
+            assert hasattr(reading, "timestamp")
+            assert hasattr(reading, "status")
+            assert hasattr(reading, "level")
+            assert hasattr(reading, "version")
+
+            # Test level properties
+            assert event.current_level is not None or event.current_level == 0
+            assert event.max_level is not None or event.max_level == 0
+            assert event.min_level is not None or event.min_level == 0
```
[the remark this answers is not in the document]
Comment (bes-dev) [reply tests/data/test_body_battery_data.py]: done

File: tests/data/test_body_battery_data.py [hunk 12]
```diff
@@ -0,0 +1,287 @@
...
+        },
+        {
+            # Invalid - missing eventStartTimeGmt
+            "event": {"eventType": "sleep"},
+            "activityName": "Test"
+        }
+    ]
+    
+    BodyBatteryData.get("2023-07-20", client=mock_client)
+    # Should process valid items and skip invalid ones
```
[the remark this answers is not in the document]
Comment (bes-dev) [reply tests/data/test_body_battery_data.py]: done

File: tests/data/test_body_battery_data.py [hunk 13]
```diff
@@ -0,0 +1,287 @@
...
+                or daily_data.current_body_battery == 0
+            )
+            assert (
+                daily_data.max_body_battery is not None
+                or daily_data.max_body_battery == 0
+            )
+            assert (
+                daily_data.min_body_battery is not None
+                or daily_data.min_body_battery == 0
+            )
```
[the remark this answers is not in the document]
Comment (bes-dev) [reply tests/data/test_body_battery_data.py]: done

**Review by matin — APPROVED** [review]

**Comment by matin** [comment]
@coderabbitai make the changes you're proposing

**Review by matin — APPROVED** [review]
